src/Core/Services/Implementations/StripePaymentService.cs

```diff
@@ -1098,6 +1083,261 @@ public async Task<string> PayInvoiceAfterSubscriptionChangeAsync(ISubscriber sub
+        }
+
+        return cardPaymentMethodId;
+    }
+
+    private async Task<Invoice> GetUpcomingInvoiceAsync(ISubscriber subscriber, List<InvoiceSubscriptionItemOptions> subItemOptions)
+    {
+        var upcomingPreview = await _stripeAdapter.InvoiceUpcomingAsync(new Stripe.UpcomingInvoiceOptions
+        {
+            Customer = subscriber.GatewayCustomerId,
+            Subscription = subscriber.GatewaySubscriptionId,
+            SubscriptionItems = subItemOptions
+        });
+        return upcomingPreview;
+    }
+
+    private async Task RestoreInvoiceItemsAsync(Invoice invoice, Customer customer, IEnumerable<InvoiceItem> pendingInvoiceItems)
+    {
+        invoice = await _stripeAdapter.InvoiceVoidInvoiceAsync(invoice.Id, new Stripe.InvoiceVoidOptions());
+        if (invoice.StartingBalance != 0)
+        {
+            await _stripeAdapter.CustomerUpdateAsync(customer.Id,
+                new Stripe.CustomerUpdateOptions { Balance = customer.Balance });
+        }
+
+        // Restore invoice items that were brought in
+        foreach (var item in pendingInvoiceItems)
+        {
+            var i = new Stripe.InvoiceItemCreateOptions
+            {
+                Currency = item.Currency,
+                Description = item.Description,
+                Customer = item.CustomerId,
+                Subscription = item.SubscriptionId,
+                Discountable = item.Discountable,
+                Metadata = item.Metadata,
+                Quantity = item.Proration ? 1 : item.Quantity,
+                UnitAmount = item.UnitAmount
+            };
+            await _stripeAdapter.InvoiceItemCreateAsync(i);
+        }
+    }
+
+    private async Task InvoicePayAsync(InvoicePayOptions invoicePayOptions, Invoice invoice, string paymentIntentClientSecret)
+    {
+        try
+        {
+            await _stripeAdapter.InvoicePayAsync(invoice.Id, invoicePayOptions);
+        }
+        catch (Stripe.StripeException e)
+        {
+            if (e.HttpStatusCode == System.Net.HttpStatusCode.PaymentRequired &&
+                e.StripeError?.Code == "invoice_payment_intent_requires_action")
+            {
+                // SCA required, get intent client secret
+                var invoiceGetOptions = new Stripe.InvoiceGetOptions();
+                invoiceGetOptions.AddExpand("payment_intent");
+                invoice = await _stripeAdapter.InvoiceGetAsync(invoice.Id, invoiceGetOptions);
+                paymentIntentClientSecret = invoice?.PaymentIntent?.ClientSecret;
+            }
+            else
+            {
+                throw new GatewayException("Unable to pay invoice.");
+            }
+        }
+    }
+
+    private async Task CreateBrainTreeTransactionRequestAsync(ISubscriber subscriber, Invoice invoice, Customer customer,
+        InvoicePayOptions invoicePayOptions, string cardPaymentMethodId, Braintree.Transaction braintreeTransaction)
+    {
+        if (invoice.AmountDue > 0)
```

Comment:
This method has a high degree of nesting. Consider inverting these conditionals to improve the readability. See below suggestion:

```C#
private async Task CreateBrainTreeTransactionRequestAsync(ISubscriber subscriber, Invoice invoice, Customer customer,
    InvoicePayOptions invoicePayOptions, string cardPaymentMethodId, Braintree.Transaction braintreeTransaction)
{
    if (invoice.AmountDue <= 0)
    {
        return;
    }

    if (!(customer?.Metadata?.ContainsKey("btCustomerId") ?? false))
    {
        invoicePayOptions.OffSession = true;
        invoicePayOptions.PaymentMethod = cardPaymentMethodId;
        return;
    }

    invoicePayOptions.PaidOutOfBand = true;
    var btInvoiceAmount = invoice.AmountDue / 100M;
    var transactionResult = await _btGateway.Transaction.SaleAsync(
        new Braintree.TransactionRequest
        {
            Amount = btInvoiceAmount,
            CustomerId = customer.Metadata["btCustomerId"],
            Options = new Braintree.TransactionOptionsRequest
            {
                SubmitForSettlement = true,
                PayPal = new Braintree.TransactionOptionsPayPalRequest
                {
                    CustomField = $"{subscriber.BraintreeIdField()}:{subscriber.Id}"
                }
            },
            CustomFields = new Dictionary<string, string>
            {
                [subscriber.BraintreeIdField()] = subscriber.Id.ToString()
            }
        });

    if (!transactionResult.IsSuccess())
    {
        throw new GatewayException("Failed to charge PayPal customer.");
    }

    braintreeTransaction = transactionResult.Target;
    await _stripeAdapter.InvoiceUpdateAsync(invoice.Id, new Stripe.InvoiceUpdateOptions
    {
        Metadata = new Dictionary<string, string>
        {
            ["btTransactionId"] = braintreeTransaction.Id,
            ["btPayPalTransactionId"] =
                braintreeTransaction.PayPalDetails.AuthorizationId
        }
    });
}
```